About parsing table names that contain backticks

## About parsing table names that contain backticks

create table ```t_order` (i int)

Hi community, currently Shardingsphere cannot parse the SQL statement that contains a table name with backticks. 
I have made changes to the corresponding parsing rules locally and now the SQL can be parsed correctly. 
However, when writing tests, I found that Shardingsphere processes the table name as "t_order" instead of "`t_order" (with backticks) after calling the constructor of IdentifierValue, which invokes SQLUtils.getExactlyValue(text). 
Is this correct behavior?


============== Test out
java.lang.AssertionError: 
SQL Case ID : create_table_with_backtick
SQL         : CREATE TABLE ```t_order` (i int)
Table name assertion error: 

Expected: is "`t_order"
     but: was "t_order"

--- a/parser/sql/dialect/mysql/src/main/antlr4/imports/mysql/BaseRule.g4
+++ b/parser/sql/dialect/mysql/src/main/antlr4/imports/mysql/BaseRule.g4
@@ -113,6 +113,7 @@ identifier
     | customKeyword
     | DOUBLE_QUOTED_TEXT
     | UNDERSCORE_CHARSET
+    | BQUOTA_STRING
     ;
     
 identifierKeywordsUnambiguous

--- a/parser/sql/dialect/mysql/src/main/antlr4/imports/mysql/Literals.g4
+++ b/parser/sql/dialect/mysql/src/main/antlr4/imports/mysql/Literals.g4
@@ -31,6 +31,10 @@ DOUBLE_QUOTED_TEXT
     : DQ_ ( '\\'. | '""' | ~('"'| '\\') )* DQ_
     ;
 
+BQUOTA_STRING
+    : BQ_ ( '\\'. | '``' | ~('`'|'\\'))* BQ_
+    ;
+
 NCHAR_TEXT
     : N SINGLE_QUOTED_TEXT
     ;

--- a/parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/util/SQLUtils.java
+++ b/parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/util/SQLUtils.java
@@ -47,13 +47,15 @@ import java.util.regex.Pattern;
 @NoArgsConstructor(access = AccessLevel.PRIVATE)
 public final class SQLUtils {
     
+    private static final char BACKTICK = '`';
+    
     private static final String SQL_END = ";";
     
     private static final String COMMENT_PREFIX = "/*";
     
     private static final String COMMENT_SUFFIX = "*/";
     
-    private static final String EXCLUDED_CHARACTERS = "[]`'\"";
+    private static final String EXCLUDED_CHARACTERS = "[]'\"";
     
     private static final Pattern SINGLE_CHARACTER_PATTERN = Pattern.compile("([^\\\\])_|^_");
     
@@ -91,14 +93,14 @@ public final class SQLUtils {
     
     /**
      * Get exactly value for SQL expression.
-     * 
+     *
      * <p>remove special char for SQL expression</p>
-     * 
+     *
      * @param value SQL expression
      * @return exactly SQL expression
      */
     public static String getExactlyValue(final String value) {
-        return null == value ? null : CharMatcher.anyOf(EXCLUDED_CHARACTERS).removeFrom(value);
+        return null == value ? null : tryGetRealContentInBackticks(CharMatcher.anyOf(EXCLUDED_CHARACTERS).removeFrom(value));
     }
     
     /**
@@ -106,7 +108,7 @@ public final class SQLUtils {
      *
      * <p>remove special char for SQL expression</p>
      *
-     * @param value SQL expression
+     * @param value              SQL expression
      * @param reservedCharacters characters to be reserved
      * @return exactly SQL expression
      */
@@ -116,6 +118,31 @@ public final class SQLUtils {
         }
         String toBeExcludedCharacters = CharMatcher.anyOf(reservedCharacters).removeFrom(EXCLUDED_CHARACTERS);
         return CharMatcher.anyOf(toBeExcludedCharacters).removeFrom(value);
+    }
+    
+    /**
+     * Try get exactly value for backticks string.
+     *
+     * <p>try get content containing backticks exactly value</p>
+     *
+     * @param value SQL expression
+     * @return exactly SQL expression
+     */
+    public static String tryGetRealContentInBackticks(final String value) {
+        if (value.charAt(0) == BACKTICK && value.charAt(value.length() - 1) == BACKTICK) {
+            char[] chars = value.toCharArray();
+            StringBuilder exactlyTableName = new StringBuilder();
+            for (int i = 1; i < chars.length - 1; i++) {
+                if (chars[i] == BACKTICK && (i + 1 >= chars.length - 1 || chars[i + 1] != BACKTICK)) {
+                    return value;
+                } else if (chars[i] == BACKTICK && chars[i + 1] == BACKTICK) {
+                    i++;
+                }
+                exactlyTableName.append(chars[i]);
+            }
+            return 0 == exactlyTableName.length() ? value : exactlyTableName.toString();
+        }
+        return value;
     }
     
     /**
@@ -132,7 +159,7 @@ public final class SQLUtils {
     
     /**
      * Get exactly SQL expression without outside parentheses.
-     * 
+     *
      * @param value SQL expression
      * @return exactly SQL expression
      */
@@ -186,11 +213,11 @@ public final class SQLUtils {
     
     /**
      * Create literal expression.
-     * 
-     * @param astNode AST node
+     *
+     * @param astNode    AST node
      * @param startIndex start index
-     * @param stopIndex stop index
-     * @param text text
+     * @param stopIndex  stop index
+     * @param text       text
      * @return literal expression segment
      */
     public static ExpressionSegment createLiteralExpression(final ASTNode astNode, final int startIndex, final int stopIndex, final String text) {
@@ -241,7 +268,7 @@ public final class SQLUtils {
     
     /**
      * Convert like pattern to regex.
-     * 
+     *
      * @param pattern like pattern
      * @return regex
      */

--- a/parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/util/SQLUtils.java
+++ b/parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/util/SQLUtils.java
@@ -47,7 +47,7 @@ import java.util.regex.Pattern;
 @NoArgsConstructor(access = AccessLevel.PRIVATE)
 public final class SQLUtils {
     
-    private static final char BACKTICK = '`';
+    private static final String BACKTICK = "`";
     
     private static final String SQL_END = ";";
     
@@ -129,16 +129,21 @@ public final class SQLUtils {
      * @return exactly SQL expression
      */
     public static String tryGetRealContentInBackticks(final String value) {
-        if (value.charAt(0) == BACKTICK && value.charAt(value.length() - 1) == BACKTICK) {
-            char[] chars = value.toCharArray();
+        if (null == value) {
+            return null;
+        }
+        if (value.startsWith(BACKTICK) && value.endsWith(BACKTICK)) {
+            int startIndex = 1;
+            int stopIndex = value.length() - 1;
             StringBuilder exactlyTableName = new StringBuilder();
-            for (int i = 1; i < chars.length - 1; i++) {
-                if (chars[i] == BACKTICK && (i + 1 >= chars.length - 1 || chars[i + 1] != BACKTICK)) {
+            while (startIndex < stopIndex) {
+                if (value.charAt(startIndex) == '`' && (startIndex + 1 >= stopIndex || value.charAt(startIndex + 1) != '`')) {
                     return value;
-                } else if (chars[i] == BACKTICK && chars[i + 1] == BACKTICK) {
-                    i++;
+                } else if (value.charAt(startIndex) == '`' && value.charAt(startIndex + 1) == '`') {
+                    startIndex++;
                 }
-                exactlyTableName.append(chars[i]);
+                exactlyTableName.append(value.charAt(startIndex));
+                startIndex++;
             }
             return 0 == exactlyTableName.length() ? value : exactlyTableName.toString();
         }

--- a/parser/sql/statement/src/test/java/org/apache/shardingsphere/sql/parser/sql/common/util/SQLUtilsTest.java
+++ b/parser/sql/statement/src/test/java/org/apache/shardingsphere/sql/parser/sql/common/util/SQLUtilsTest.java
@@ -63,6 +63,8 @@ class SQLUtilsTest {
         assertThat(SQLUtils.getExactlyValue("[xxx]"), is("xxx"));
         assertThat(SQLUtils.getExactlyValue("\"xxx\""), is("xxx"));
         assertThat(SQLUtils.getExactlyValue("'xxx'"), is("xxx"));
+        assertThat(SQLUtils.getExactlyValue("`[xxx`"), is("xxx"));
+        assertThat(SQLUtils.getExactlyValue("```[xxx```"), is("`xxx`"));
     }
     
     @Test
@@ -152,5 +154,19 @@ class SQLUtilsTest {
         assertThat(SQLUtils.trimComment("/* This is a comment */ SHOW DATABASES"), is("SHOW DATABASES"));
         assertThat(SQLUtils.trimComment("/* This is a query with a semicolon */ SHOW DATABASES;"), is("SHOW DATABASES"));
         assertThat(SQLUtils.trimComment("/* This is a query with spaces */    SHOW DATABASES   "), is("SHOW DATABASES"));
+    }
+    
+    @Test
+    void assertTryGetRealContentInBackticks() {
+        assertThat(SQLUtils.tryGetRealContentInBackticks("`"), is("`"));
+        assertThat(SQLUtils.tryGetRealContentInBackticks("``"), is("``"));
+        assertThat(SQLUtils.tryGetRealContentInBackticks("```"), is("```"));
+        assertThat(SQLUtils.tryGetRealContentInBackticks("t_order"), is("t_order"));
+        assertThat(SQLUtils.tryGetRealContentInBackticks("````"), is("`"));
+        assertThat(SQLUtils.tryGetRealContentInBackticks("`t_order`"), is("t_order"));
+        assertThat(SQLUtils.tryGetRealContentInBackticks("```t_order`"), is("`t_order"));
+        assertThat(SQLUtils.tryGetRealContentInBackticks("```````t_order```````"), is("```t_order```"));
+        assertThat(SQLUtils.tryGetRealContentInBackticks("``t_order`"), is("``t_order`"));
+        assertThat(SQLUtils.tryGetRealContentInBackticks("`````t_o``r``d``e``r```"), is("``t_o`r`d`e`r`"));
     }
 }

--- a/test/it/parser/src/main/resources/case/ddl/create-table.xml
+++ b/test/it/parser/src/main/resources/case/ddl/create-table.xml
@@ -17,6 +17,12 @@
   -->
 
 <sql-parser-test-cases>
+    <create-table sql-case-id="create_table_with_backtick">
+        <table name="t_order" start-delimiter="`" end-delimiter="`" start-index="13" stop-index="23" />
+        <column-definition type="int" start-index="26" stop-index="30">
+            <column name="i" />
+        </column-definition>
+    </create-table>
     <create-table sql-case-id="create_table_with_like">
         <table name="t_log" start-index="13" stop-index="17" />
         <like-table name="t_old_log" start-index="24" stop-index="32" />

--- a/test/it/parser/src/main/resources/case/ddl/create-table.xml
+++ b/test/it/parser/src/main/resources/case/ddl/create-table.xml
@@ -18,9 +18,15 @@
 
 <sql-parser-test-cases>
     <create-table sql-case-id="create_table_with_backtick">
-        <table name="t_order" start-delimiter="`" end-delimiter="`" start-index="13" stop-index="23" />
+        <table name="`t_order" start-delimiter="`" end-delimiter="`" start-index="13" stop-index="23" />
         <column-definition type="int" start-index="26" stop-index="30">
             <column name="i" />
+        </column-definition>
+    </create-table>
+    <create-table sql-case-id="create_table_column_with_backtick">
+        <table name="``t_o`r`d`e`r`" start-delimiter="`" end-delimiter="`" start-index="13" stop-index="35" />
+        <column-definition type="int" start-delimiter="`" end-delimiter="`" start-index="38" stop-index="46">
+            <column name="`i" />
         </column-definition>
     </create-table>
     <create-table sql-case-id="create_table_with_like">

--- a/test/it/parser/src/main/resources/sql/supported/ddl/create-table.xml
+++ b/test/it/parser/src/main/resources/sql/supported/ddl/create-table.xml
@@ -17,6 +17,7 @@
   -->
 
 <sql-cases>
+    <sql-case id="create_table_with_backtick" value="CREATE TABLE ```t_order` (i int)" db-types="MySQL" />
     <sql-case id="create_table_with_like" value="CREATE TABLE t_log LIKE t_old_log" db-types="MySQL" />
     <sql-case id="create_table" value="CREATE TABLE t_log(id int PRIMARY KEY, status varchar(10))" />
     <sql-case id="create_table_with_engin_charset" value="CREATE TABLE t_log(id int PRIMARY KEY, status varchar(10)) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4" db-types="MySQL" />

--- a/test/it/parser/src/main/resources/sql/supported/ddl/create-table.xml
+++ b/test/it/parser/src/main/resources/sql/supported/ddl/create-table.xml
@@ -18,6 +18,7 @@
 
 <sql-cases>
     <sql-case id="create_table_with_backtick" value="CREATE TABLE ```t_order` (i int)" db-types="MySQL" />
+    <sql-case id="create_table_column_with_backtick" value="create table `````t_o``r``d``e``r``` (```i` int)" db-types="MySQL" />
     <sql-case id="create_table_with_like" value="CREATE TABLE t_log LIKE t_old_log" db-types="MySQL" />
     <sql-case id="create_table" value="CREATE TABLE t_log(id int PRIMARY KEY, status varchar(10))" />
     <sql-case id="create_table_with_engin_charset" value="CREATE TABLE t_log(id int PRIMARY KEY, status varchar(10)) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4" db-types="MySQL" />

